### DISCUSSION TITLE
Receive MIDI Inputs in ableton and napkin application

### DISCUSSION THREAD

**Discussion by jkukatzki** [napframework/nap#35] [Q&A]
Hello again, I'm building a small interactive installation where you can play some instruments put together in ableton and have some visuals play to it on a CRT screen. My initial plan was to not do this using audio analysis but rather just take the input from the same MIDI Keyboard that controls ableton, but I cannot get this to work as either ableton or my nap app seem to hog the device once the program starts and can't be used by them at the same time.
Hoping to get some advice on this because I'd be a lot cooler than just using audio reactive stuff. I tried the MidiOutputPort resource but am a bit lost on how it works and I lack some knowledge about MIDI I think and what a port is actually defined by and where its name comes from. Maybe someone else has already solved this.

Thanks!

**Comment by cklosters** [comment]
Hey, 

I know of many that use NAP in combination with Ableton & Midi / OSC, but not me personally. Maybe @stijnvanbeek can help you out here, but we would need additional info (hogging system?). Midi is just midi, there's nothing much to it - you send or receive midi over midi ports. There's an `oscmidi` demo that's worth looking at, also: consider using OSC instead of MIDI, higher resolution and more control.

**Comment by stijnvanbeek** [comment]
Hi there,

You need a virtual midi driver to rout the midi messages from Ableton to nap or vice versa. If you are on Mac OS you can use the midi IAC driver for that purpose, which is part of the os. I would advice to rout from Ableton to nap. Then you can use the midi input port resource combined with a Midi input component. Hope this helps! Best,
Stijn
